chore: add ADR for OSI token (#2874)
opensafely-core/job-server: docs/adr/0014-osi-token.md

# 14. Github token set up for OSI
Date: 2023-03-24

## Status
Accepted

## Context

For OSI, job-server needs to be able to write generated code to Github.

> Previously, job-server only had read access to repos, as a security control.

Author: I mean, yes, but the design is that is was readonly. The fact that there's a loop hole I don't think is really relevant to this decision or consequences.

And fixing that loop hole will also probably require the secure service.

> Previously, job-server only had read access to repos, as a security control.

Author: There's a small window where *write access to a brand new repo* is possible, for clarity. Only repo creation is the problem.

Granting arbitrary repo write permissions to job-server weakens the security

> layering, as it increases the possibly attcak vectors of what compromising

Reviewer: ```suggestion
layering, as it increases the possible attack vectors of what compromising
```

job-server alone can do.

The current read-only access is provided by the opensafely-readonly bot

> account, as classic PATs can not be restricted to read only access in any other

Reviewer: ```suggestion
account, as classic PATs cannot be restricted to read only access in any other
```

way.

We explored the new fine-grained PATs, but you can only restrict access to an
explict list of repos, which mean we'd need to update or regenerate the token
every time we added a new interactive study.

## Decision

We created an [Interactive
team](https://github.com/orgs/opensafely/teams/interactive/members) in the
opensafely org. We added the pre-existing `opensafely-interactive-bot` as
a member of this team.

When we create a new interactive repo, we automatically add it to this team
with `write` permissions. This means a classic PAT generated by the
opensfely-interactive-bot user can only write to those repos.

## Consequences

Job-server does now have a token than can write to repos. It is at least
restricted to writing to interactive repos, and cannot access user's repos

This unblocks OSIv2, but it does however remove the layer of protection we had
originaly designed.  A compromised job-server can now run arbitrary code in any
backend. We still have our other layers of protection (Level 4 access

> controls), so this is not critical.

Author: Compromising job-server does bypass our output checking process, as we use job-server to know who is allowed to release.
